File: merge_god/send_approval.py
#!/usr/bin/env python3
"""
Quick script to send approval to a waiting pr-loop.py process
"""

import json
import sys
from pathlib import Path

import psutil


def find_pr_loop_process():
    """Find the running pr-loop.py process"""
    for proc in psutil.process_iter(["pid", "name", "cmdline"]):
        try:
            cmdline = proc.info.get("cmdline", [])
            if cmdline and "pr-loop.py" in " ".join(cmdline):
                return proc.info["pid"]
        except (psutil.NoSuchProcess, psutil.AccessDenied):
            continue
    return None


def send_approval(pid):
    """Send approval JSON to process stdin"""
    try:
        # Open the stdin fd of the process
        stdin_path = f"/proc/{pid}/fd/0"
        approval = {"approved": True}

        # Try to write to stdin
        with Path(stdin_path).open("w") as stdin:
            stdin.write(json.dumps(approval) + "\n")
            stdin.flush()

        print(f"✓ Sent approval to pr-loop.py (PID {pid})")
        return True
    except Exception as e:
        print(f"✗ Failed to send approval: {e}", file=sys.stderr)
        return False


def main():
    print("Looking for pr-loop.py process...")
    pid = find_pr_loop_process()

    if not pid:
        print("✗ No pr-loop.py process found", file=sys.stderr)
        sys.exit(1)

    print(f"Found pr-loop.py process (PID {pid})")
    print("Sending approval...")

    if send_approval(pid):
        sys.exit(0)
    else:
        sys.exit(1)


if __name__ == "__main__":
    main()

File: send_approval.ts
/**
 * Quick script to send approval to a waiting pr-loop process.
 *
 * Ported from send_approval.py. Finds the running pr-loop process by
 * scanning /proc, then writes a `{"approved": true}` JSON line to that
 * process's stdin (/proc/{pid}/fd/0).
 */

import { readFileSync, readdirSync, writeFileSync } from "node:fs";
import { pathToFileURL } from "node:url";

/**
 * Find the running pr-loop process.
 *
 * Mirrors psutil.process_iter(["pid", "name", "cmdline"]): iterates the numeric
 * entries under /proc, reads each cmdline file (NUL-separated argv joined by
 * spaces), and returns the first PID whose cmdline contains "pr-loop".
 * Processes that vanish or are inaccessible are skipped, matching
 * NoSuchProcess/AccessDenied handling. Returns null when no match is found.
 */
export function findPrLoopProcess(): number | null {
  let entries: string[];
  try {
    entries = readdirSync("/proc");
  } catch {
    return null;
  }
  for (const entry of entries) {
    if (!/^\d+$/.test(entry)) continue;
    let cmdline: string;
    try {
      const raw = readFileSync(`/proc/${entry}/cmdline`, "utf8");
      cmdline = raw.replace(/\0+$/g, "").replace(/\0/g, " ");
    } catch {
      continue;
    }
    if (cmdline.includes("pr-loop.ts") || cmdline.includes("merge-god-pr-loop")) {
      return Number(entry);
    }
  }
  return null;
}

/**
 * Send approval JSON to a process's stdin.
 *
 * Writes `{"approved": true}\n` to /proc/{pid}/fd/0 (truncate + write, matching
 * Python's `open(path, "w")`). Prints a success message and returns true on
 * success; prints the failure to stderr and returns false otherwise.
 */
export function sendApproval(pid: number): boolean {
  try {
    const stdinPath = `/proc/${pid}/fd/0`;
    const approval = { approved: true };
    writeFileSync(stdinPath, JSON.stringify(approval) + "\n", "utf8");
    console.log(`✓ Sent approval to pr-loop (PID ${pid})`);
    return true;
  } catch (e) {
    console.error(`✗ Failed to send approval: ${e instanceof Error ? e.message : String(e)}`);
    return false;
  }
}

function main(): void {
  console.log("Looking for pr-loop process...");
  const pid = findPrLoopProcess();

  if (!pid) {
    console.error("✗ No pr-loop process found");
    process.exit(1);
  }

  console.log(`Found pr-loop process (PID ${pid})`);
  console.log("Sending approval...");

  if (sendApproval(pid)) {
    process.exit(0);
  } else {
    process.exit(1);
  }
}

if (import.meta.url === pathToFileURL(process.argv[1] ?? "").href) {
  main();
}
